text/0000-iri-v181-transaction-and-bundle/0000-iri-v181-transaction-and-bundle.md

```diff
@@ -0,0 +1,980 @@
+    ///
+    /// NOTE: `iri v1.8.1` uses `Kerl` as a sponge.
+    ///
+    /// NOTE: This function is defined on: `SealedOutgoingBundleBuilder`, `OutgoingBundleBuilder`, and `IncomingBundleBuilder`.
+    ///
+    /// TODO: This function relies on some `Sponge` trait, which is not yet defined. The code below lays out
+    ///       how the sponge is thought to be used.
+    pub fn calculate_hash<S: Sponge<Output=BundleHash>>(&self, mut sponge: S) -> S::Output {
+        unimplemented!();
+
+        // Pseudocode of how `calculate_hash` should look.
+
+        for transaction_draft in bundle_builder {
+            sponge.absorb(transaction_draft.essence());
+        }
+        sponge.squeeze()
+    }
+
+    /// Chains the contained transactions, filling the `nonce` fields via proof of work and then hashing transactions
+    /// to reference successive transactions.
+    ///
+    /// Using some type `P: ProofOfWork`, this function peforms proof of work and updates the `nonce` field in each of
+    /// the contained `TransactionBuilder`s. The transactions' hashes can then be calculated, and a chain is established
+    /// between transactions by putting the hash of a transaction into its downstream neighbor.
+    ///
+    /// Proof of Work (PoW) allows your transactions to be accepted by the network. On the IOTA network, PoW is only
+    /// a rate control mechanism. After PoW, a bundle is ready to be sent to the network. Given a `trunk` and
+    /// a `branch` returned by [`getTransactionsToApprove`], the process iterates over each transaction of the bundle,
+    /// sets some attachment related fields and does individual PoW.
+    ///
+    /// [`getTransactionsToApprove`]: https://docs.iota.org/docs/node-software/0.1/iri/references/api-reference#gettransactionstoapprove
+    ///
+    /// TODO: This function relies on some `Sponge` and `ProofOfWork` traits, which are not yet defined. The code below lays out
+    ///       how proof of work is thought to be used.
+    pub fn chain_and_attach<S: Sponge, P: ProofOfWork>(
+        mut self,
+        trunk_tip: TransactionHash,
+        branch_tip: TransactionHash,
+        proof_of_work: P,
+        sponge: S,
+    ) -> Result<Self, OutgoingBundleError> {
+        unimplemented!()
+
+        // Iterates over all transactions starting from the head, setting trunks, branches, timestamps, tags, and nonce.
+        //
+        // NOTE: We do not provide an implementation here.
+        let builder_iterator = self.iter_from_head();
+
+        // Set the head transaction
+        let head_transaction = builder_iterator.next().ok_or("builder does not contain any transaction builders")?;
+        head_transaction.trunk(trunk_tip);
+        head_transaction.branch(branch_tip);
+
+        head_transaction.attachment_timestamp = timestamp();
+        head_transaction.attachment_timestamp_lower = 0;
+        head_transaction.attachment_timestamp_upper = 3812798742493;
```

Comment:
It's a little bit confusing that no explanation about this magic number here.

Reply:
Added a `FIXME` to the code because I am not sure where that number comes from.